Liveness and UpperVector Save/Restore

kunalspathak opened the pull request:
As I pointed out in https://github.com/dotnet/runtime/issues/98718#issuecomment-1967475467, we falsely think that a variable is not live anymore and avoid inserting upper vector save/restore for that variable around a call. 

Today, we insert upper vector save/restore only if vector registers are live across the given call, however that information might be inaccurate. To address this, we will now try to take into account all the live-in and new definition variables that gets created in the given block and insert upper vector save/restore for them. We cannot eliminate the variables that are not liveOut, because we might end up having a sequence where we are using a variable (live-in or new definition) after the call and hence they need to be saved and restored. Due to this, we will end up creating more RefPositions than what we do currently, but hopefully it shouldn't cause much TP impact (need to check from the CI).

There is one case we need to handle where the definition of a variable can appear after the call:

```
...
CALL
...
V12 = 
...
 = V12
```

For these cases , today, we will insert upper save/restore around the CALL, but when we see that it gets defined after the call before a use, we reset the information and ignore the save/restore for them (by setting `isPartiallySpilled = false`). With this PR, we will end up bunch of save/restore and we can run into this situation and we hit assert at the block boundary, when we verify that none of the vector variables are still marked as partially spilled because we expect that we have seen an equivalent UpperVectorRestore in the block. It basically verifies that a pair of UpperVectorSave/UpperVectorRestore exists in a given block , which holds true if the variable is *liveVars*. However, with the change in this PR, we might have few extra `UpperVectorSave` without UpperVectorRestore or 2 `UpperVectorSave` back to back. To mitigate that, I have added a check to ignore the *extra* `UpperVectorSave` that was added and do not allocate register for it. 

The liveness handling should ideally be done before LSRA and we shouldn't be doing the special handling we have to in this PR for UpperVectorSave/UpperVectorRestore. Additionally, we are just increasing the amount of RefPositions which would increase TP for methods containing vector registers.

Alternatively, today our IR model loads the `LCL_VAR` at any place and it only gets consumed at the user. Hence the gap between the load and use of `LCL_VAR` might introduce such liveness holes, leading to functional correctness bug. I am surprised that we didn't see this issue until now. But, it will be lot of work to fix the IR model or rearranging the `LCL_VAR` placement.

 
Fixes: #98718

kunalspathak committed:
take the def/live-in variables into account for upper vector save/restore

kunalspathak commented:
Seems to be very pervasive change with various asserts hitting. Will investigate.

```
[03:09:19] ISSUE: <ASSERT> #44281 D:\a\_work\1\s\src\coreclr\jit\lsra.cpp (7987) - Assertion failed '(genRegMask(localVarInterval->physReg) & getKillSetForNode(treeNode)) == RBM_NONE' in 'Algorithms.VectorFloatRenderer:RenderSingleThreadedWithADT(float,float,float,float,float):this' during 'LSRA allocate' (IL size 421; hash 0xfb484978; Tier1-OSR)

[03:09:19] 

[03:09:19] ISSUE: <ASSERT> #81781 D:\a\_work\1\s\src\coreclr\jit\simdcodegenxarch.cpp (443) - Assertion failed 'varDsc->lvOnFrame' in 'System.PackedSpanHelpers:IndexOf[System.SpanHelpers+DontNegate`1[short],System.PackedSpanHelpers+Or20Transform](byref,short,int):int' during 'Generate code' (IL size 1081; hash 0x7e3d3108; Tier1)

[03:09:19] 

[03:09:19] ISSUE: <ASSERT> #43962 D:\a\_work\1\s\src\coreclr\jit\lsra.cpp (4716) - Assertion failed '!Compiler::varTypeNeedsPartialCalleeSave(interval->registerType) || !interval->isPartiallySpilled' in 'Algorithms.VectorFloatRenderer:RenderSingleThreadedNoADT(float,float,float,float,float):this' during 'LSRA allocate' (IL size 493; hash 0x6eaff07b; Tier1-OSR)

[
```

kunalspathak committed:
skip saveRestore and marking partially skipped for extra save refposition that we creat

kunalspathak committed:
jit format

kunalspathak commented:
> Seems to be very pervasive change with various asserts hitting

It was not that bad. I forgot to ignore the extra `RefTypeUpperVectorSave` we added.

tannergooding commented:
It looks like for x64, there's a lot of locals having their stack offset shifted, even though no new stack locals actually exist. This is pushing many variables outside the `[-128, +127]` offset range and causing the encoding to increase in size.

Ideally we could support this without unnecessarily shifting stack offsets since no new stack data is actually being placed in those slots.

kunalspathak committed:
fix the assert

kunalspathak committed:
Add liveVarUpperSave

`liveVarUpperSave` will give a hint to the allocator that the given refposition
is something that was added conservatively and might not be live. Allocator when
sees this refposition and if the corresponding lclVarInterval was not assigned a
register, it will skip assiging or deallocating register for the current Refposition
as well. That way, the end effect of the extra RefPositions added for non-liveVars
is reduced to minimum.

kunalspathak committed:
jit format and comments

kunalspathak committed:
fix the release build

kunalspathak commented:
> It looks like for x64, there's a lot of locals having their stack offset shifted

Yes, I noticed that too and added an optimization where we will not spill/allocate for non-liveVar upperSave if their corresponding lclVarInterval are not yet assigned a register. Additionally, when we come across such intervals, we will also unassign the register so it can be used for different variable. Locally I saw good codesize gains because we can reuse the callee-save registers and hence the extra code for saving/restoring them in prolog/epilog is gone.

kunalspathak committed:
fix formatting

tannergooding commented on src/coreclr/jit/lsra.cpp:
```diff
@@ -5922,9 +5922,62 @@ void LinearScan::allocateRegisters()
-                if ((lclVarInterval->physReg == REG_NA) ||
+                assert(currentInterval->recentRefPosition == &currentRefPosition);
+
+                // For a given RefTypeUpperVectorSave, there should be a matching RefTypeUpperVectorRestore
+                // If not, probably, this was an extra one we added conservatively and should not need a register.
```
I don't understand this comment.

If we're saving something across a callsite, we must restore it. Otherwise, the save was never needed in the first place and it can simply be bypassed.

kunalspathak replied:
> Otherwise, the save was never needed in the first place and it can simply be bypassed.

That's the whole point. Basically, earlier we were saving only for live variables, but we found out that sometimes the liveness information can be inaccurate and so we are saving for all the `bbLiveIn` and `bbDefs`, but later, we try to find out and see which one of the saves are unnecessary and bypass it by the virtue of this and `IsExtraUpperVectorSave`

tannergooding replied:
Can we not just fix liveness to ensure that other presumptions aren't incorrect?

Seemingly if liveness is wrong, any other phase of the JIT could also be doing things incorrectly.

kunalspathak commented:
/azp run runtime-coreclr libraries-jitstress2-jitstressregs, runtime-coreclr libraries-jitstressregs, 
runtime-coreclr jitstressregs

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into uppervector-liveness

jakobbotsch commented on src/coreclr/jit/lsrabuild.cpp:
```diff
@@ -1585,25 +1609,36 @@ void LinearScan::buildUpperVectorRestoreRefPosition(
+        unsigned     varIndex              = lclVarInterval->getVarIndex(compiler);
+        Interval*    upperVectorInterval   = getUpperVectorInterval(varIndex);
+        RefPosition* savePos               = upperVectorInterval->recentRefPosition;
+        if (!isUse && !savePos->liveVarUpperSave)
+        {
+            // If we are just restoring upper vector at the block boundary and if this is not
+            // a upperVector related to the liveVar, then ignore creating restore for them.
+            // During allocation, we will detect that this was an extra save-upper and skip
+            // the save/restore altogether.
+            return;
+        }
```
What is this case about? How do we know this is at a block boundary here?

kunalspathak replied:
We pass `isUse = false` at the block boundary. If they were not live variables i.e. `!liveVarUpperSave`, we know that these are the extra ones that we added and are not really needed. Had they were needed, we would have already built `RefUpperVectorRestore` at their use before we reached the block boundary.

jakobbotsch commented on src/coreclr/jit/lsrabuild.cpp:
```diff
@@ -1508,6 +1528,10 @@ void LinearScan::buildUpperVectorSaveRefPositions(GenTree* tree, LsraLocation cu
                     newRefPosition(upperVectorInterval, currentLoc, RefTypeUpperVectorSave, tree, RBM_FLT_CALLEE_SAVED);
                 varInterval->isPartiallySpilled = true;
                 pos->skipSaveRestore            = blockAlwaysReturn;
+                if (VarSetOps::IsMember(compiler, liveLargeVectors, varIndex))
+                {
+                    pos->liveVarUpperSave = true;
+                }
```
Nit:
```suggestion
                pos->liveVarUpperSave = VarSetOps::IsMember(compiler, liveLargeVectors, varIndex);
```
?

jakobbotsch commented on src/coreclr/jit/lsrabuild.cpp:
```diff
@@ -1491,12 +1492,31 @@ void LinearScan::buildUpperVectorSaveRefPositions(GenTree* tree, LsraLocation cu
+        VARSET_TP liveLargeVectors(VarSetOps::Intersection(compiler, currentLiveVars, largeVectorVars));
+
+#ifdef DEBUG
+        assert(VarSetOps::IsEmpty(compiler, liveLargeVectors) ||
+               VarSetOps::IsSubset(compiler, liveLargeVectors, liveDefsLargeVectors));
```
Nit:
```suggestion
        assert(VarSetOps::IsSubset(compiler, liveLargeVectors, liveDefsLargeVectors));
```
If it's empty it should also be a subset

jakobbotsch replied:
(Can also remove the #ifdef DEBUG)

kunalspathak replied:
> (Can also remove the #ifdef DEBUG)

yes, I had a code earlier inside this, but i removed it later. So yes, `#ifdef DEBUG` is not needed anymore.

> If it's empty it should also be a subset

true, not sure at one point I was hitting an assert. May be my calculation were wrong. Removed it.

jakobbotsch commented on src/coreclr/jit/lsrabuild.cpp:
```diff
@@ -1508,6 +1528,10 @@ void LinearScan::buildUpperVectorSaveRefPositions(GenTree* tree, LsraLocation cu
                     newRefPosition(upperVectorInterval, currentLoc, RefTypeUpperVectorSave, tree, RBM_FLT_CALLEE_SAVED);
                 varInterval->isPartiallySpilled = true;
                 pos->skipSaveRestore            = blockAlwaysReturn;
```
This is preexisting, but why is it ok to ignore save/restore in a block that always returns? How does that work if there are more uses of the vector local after the call?

kunalspathak replied:
yes, I think there was an example I was hitting this, but may be because I was not handling the upper vector save/restore correctly at other places. So this should not be needed. I will remove it.

jakobbotsch replied:
Ah, I think I understand now... `skipSaveRestore` gets reset again by `buildUpperVectorRestoreRefPosition` if we end up seeing a use. So I think having `BBJ_RETURN` in the check there would be ok.

jakobbotsch reviewed (approved):
LGTM, just had a few nits and some questions.

kunalspathak committed:
address feedback

kunalspathak commented:
From the [diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=589832&view=ms.vss-build-web.run-extensions-tab):

![image](https://github.com/dotnet/runtime/assets/12488060/579f51eb-1773-43f5-9a64-fd89b1da0512)
